File: 3291-find-if-array-can-be-sorted/find-if-array-can-be-sorted.py
class Solution:
    def countSetBits(self, n: int) -> int:
        count = 0
        while n != 0:
            n = n & (n - 1)
            count += 1
        return count
    """
    Approach
        1. Split the array into segments. Each segment contains 
           consecutive elements with the same number of set bits.
        2. From left to right, the previous segment’s largest 
           element should be smaller than the current segment’s smallest element.

    Time complexity: O(m * log n), 
    where n is the value of the largest number in the array
    and we have m elements in nums.
    Space complexity: O(m), where m is the size of the input array nums.
    """
    def canSortArray(self, nums: List[int]) -> bool:
        segments = [] 
        prev_setbitcount = -1

        for num in nums:
            cur_setbitcount = self.countSetBits(num)
            if cur_setbitcount == prev_setbitcount:
                segments[-1].append(num)
            else:
                segments.append([num])
                prev_setbitcount = cur_setbitcount
        
        for i in range(len(segments) - 1):
            if max(segments[i]) > min(segments[i + 1]):
                return False
        return True

File: 3291-find-if-array-can-be-sorted/find-if-array-can-be-sorted.cpp
class Solution {
public:
    // Brian Kernighan's Algorithm to 
    // count the number of set bits (1s) in an integer's binary representation.
    // Time Complexity: O(log n)
    // Space Complexity: O(1)
    int countSetBits(int n) {
        int count = 0;
        while (n != 0) {
            n = n & (n - 1);
            count++;
        }
        return count;
    }
    // Time complexity: O(m * log n), 
    // where n is the value of the largest number in the array
    // and we have m elements in nums.
    // Space complexity: O(m), where m is the size of the input array nums.
    bool canSortArray(vector<int>& nums) {
        vector<vector<int>>segments;
        int cur_setbitcount = 0;
        int prev_setbitcount = 0;
        
        for(int i = 0; i < nums.size(); i++) {
            cur_setbitcount = countSetBits(nums[i]);
            if (prev_setbitcount == cur_setbitcount) {
                segments.back().push_back(nums[i]);
            } else {
                vector<int> segment;
                segment.push_back(nums[i]);
                segments.push_back(segment);
                prev_setbitcount = cur_setbitcount;   
            }
        }
        for (int i = 0 ; i < segments.size() - 1; i++) {
            if (*max_element(segments[i].begin(),segments[i].end())
            > *min_element(segments[i+1].begin(),segments[i+1].end())) {
                return false;
            }
        }
        return true;
    }
};